### PR TITLE
Issue 2902: Prevent NPE when truncate leaves trailing bytes in the segment

### LINKED ISSUE #2902: Stream Truncation results in a segment with length 1 byte which is incorrect
**Stream Truncation results in a segment with length 1 byte which is incorrect**

Longevity run failed with reader throws a TruncatedDataException after 6hrs.
Experiments has been executed in a virtualized platform. The cluster contains 5 nodes (CentOS 7 VMs) that run the different Pravega services in containers (Mesos): 2 Master node and 3 slaves Continuous smallScale read / write workload (3 reader and 3 writer).

**Problem location**
segment store/client ?

```
StreamSegmentInfo for this segment (from a prev invocation)
2018-09-05 17:55:23,533 21608185 [epollEventLoopGroup-5-23] DEBUG i.p.client.netty.impl.RawClient - Received reply WireCommands.StreamSegmentInfo(type=STREAM_SEGMENT_INFO, requestId=1, segmentName=hulk/smallScale/333.#epoch.207, exists=true, isSealed=true, isDeleted=false, lastModified=1536170123019, writeOffset=1766330, startOffset=1766329)

2018-09-05 17:55:24,150 21608802 [pool-8-thread-4] ERROR c.e.n.h.t.w.readers.PravegaReader - Got a truncated data exception on forgetful reader HulkSmallScale82c7d445469342f59defbf687a17e442 after position hulk/smallScale/360.#epoch.224=0, hulk/smallScale/342.#epoch.213=224388, hulk/smallScale/329.#epoch.204=1043882, hulk/smallScale/322.#epoch.200=1320061, hulk/smallScale/332.#epoch.206=822326, hulk/smallScale/331.#epoch.206=867424, hulk/smallScale/346.#epoch.215=438575, hulk/smallScale/362.#epoch.226=0, hulk/smallScale/335.#epoch.208=1640997, hulk/smallScale/341.#epoch.212=0, hulk/smallScale/343.#epoch.213=247371, hulk/smallScale/316.#epoch.196=1780403, hulk/smallScale/333.#epoch.207=1766329, hulk/smallScale/315.#epoch.196=1733322, hulk/smallScale/345.#epoch.215=0, hulk/smallScale/313.#epoch.194=1730356, hulk/smallScale/337.#epoch.209=1682542, hulk/smallScale/328.#epoch.204=1037030
io.pravega.client.stream.TruncatedDataException: null
    at io.pravega.client.stream.impl.EventStreamReaderImpl.handleSegmentTruncated(EventStreamReaderImpl.java:235)
    at io.pravega.client.stream.impl.EventStreamReaderImpl.readNextEvent(EventStreamReaderImpl.java:110)
    at com.emc.nautilus.hulk.taskexecutor.workers.readers.PravegaReader.start(PravegaReader.java:114)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
    at java.lang.Thread.run(Unknown Source)
2018-09-05 17:55:24,151 21608803 [epollEventLoopGroup-5-12] INFO  i.p.client.netty.impl.RawClient - Closing connection with exception: null
2018-09-05 17:55:24,151 21608803 [epollEventLoopGroup-5-6] DEBUG i.p.c.n.i.ClientConnectionInboundHandler - 172.16.19.23 processing reply: WireCommands.SegmentRead(type=SEGMENT_READ, segment=hulk/smallScale/333.#epoch.207, offset=1766330, atTail=false, endOfSegment=true, data=java.nio.HeapByteBuffer[pos=0 lim=0 cap=0])
2018-09-05 17:55:24,151 21608803 [pool-8-thread-4] ERROR c.e.n.h.t.w.readers.PravegaReader - Reader finished with Error
java.lang.NullPointerException: null
    at io.pravega.common.concurrent.Futures.lambda$await$0(Futures.java:68)
    at io.pravega.common.Exceptions.handleInterrupted(Exceptions.java:90)
    at io.pravega.common.concurrent.Futures.await(Futures.java:66)

java.lang.NullPointerException
    at io.pravega.common.concurrent.Futures.lambda$await$0(Futures.java:68)
    at io.pravega.common.Exceptions.handleInterrupted(Exceptions.java:90)
    at io.pravega.common.concurrent.Futures.await(Futures.java:66)
    at io.pravega.client.segment.impl.SegmentInputStreamImpl.readEventData(SegmentInputStreamImpl.java:137)
    at io.pravega.client.segment.impl.SegmentInputStreamImpl.read(SegmentInputStreamImpl.java:111)
    at io.pravega.client.stream.impl.EventStreamReaderImpl.readNextEvent(EventStreamReaderImpl.java:104)
    at com.emc.nautilus.hulk.taskexecutor.workers.readers.PravegaReader.start(PravegaReader.java:114)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
    at java.lang.Thread.run(Unknown Source)

[epollEventLoopGroup-3-26] ERROR i.p.s.s.h.h.ServerConnectionInboundHandler - Caught exception on connection
```

[Mr Hyde - Longevity log.zip](https://github.com/pravega/pravega/files/2370722/Mr.Hyde.-.Longevity.log.zip)

### PR DESCRIPTION
**Pull request by tkaitchuck**
**Change log description**  
* Skip over trailing bytes at the end of a segment if they are smaller than the length of an event header.

**Purpose of the change**  
Fixes #2902. But does not actually identify why the truncate was leaving a single trailing byte.

**What the code does**  
1. Removes a redundant parameter to truncateSegment().
2. Adds a truncation integration test.
3. Skips over trailing bytes at the end of the segment if they can't form a full event.
4. Adds a test for this case.

### CODE REVIEW

File: client/src/main/java/io/pravega/client/segment/impl/SegmentMetadataClient.java [hunk 1]
```diff
@@ -49,10 +49,9 @@
      * Deletes all data before the offset of the provided segment.
      * This data will no longer be readable. Existing offsets are not affected by this operations. 
      * The new startingOffset will be reflected in {@link SegmentMetadataClient#getSegmentInfo(String).startingOffset}.
-     * @param segment The segment to truncate.
```
Comment (shrids) [inline client/src/main/java/io/pravega/client/segment/impl/SegmentMetadataClient.java]: The javadoc link to getSegmentInfo has an incorrect input parameter.

File: test/integration/src/test/java/io/pravega/test/integration/TruncateTest.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        ack = producer.writeEvent(testString);
+        ack.get(5, TimeUnit.SECONDS);
+        
+        streamManager.createReaderGroup("ReaderGroup", ReaderGroupConfig.builder().disableAutomaticCheckpoints().stream(new StreamImpl(SCOPE, streamName)).build());
+        @Cleanup
+        EventStreamReader<String> reader = clientFactory.createReader("reader", "ReaderGroup", serializer, ReaderConfig.builder().build());
+        AssertExtensions.assertThrows(TruncatedDataException.class, () -> reader.readNextEvent(2000));
+        EventRead<String> event = reader.readNextEvent(2000);
+        assertEquals(testString, event.getEvent());
+        event = reader.readNextEvent(100);
```
Comment (fpj) [inline test/integration/src/test/java/io/pravega/test/integration/TruncateTest.java]: This is not actually testing the NPE path as the truncation is happening at the write offset, leaving the segment empty. Of course, it relies on the truncation doing the right thing, but I think that to test that we don't throw in the presence of an imprecise truncation, we would need to leave at least one byte in the segment.

File: test/integration/src/test/java/io/pravega/test/integration/TruncateTest.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+import java.util.concurrent.TimeoutException;
+import lombok.Cleanup;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+public class TruncateTest {
```
Comment (fpj) [inline test/integration/src/test/java/io/pravega/test/integration/TruncateTest.java]: We test truncation at various places in the set of test cases of the client. Have you thought either having this test case in a different class or consolidating other tests that perform truncation in a single test class?

File: client/src/main/java/io/pravega/client/segment/impl/SegmentMetadataClient.java [hunk 4]
```diff
@@ -49,10 +49,9 @@
...
      * @param offset The offset the segment should be truncated at.
      */
-    abstract void truncateSegment(Segment segment, long offset);
+    abstract void truncateSegment(long offset);
```
Comment (fpj) [inline client/src/main/java/io/pravega/client/segment/impl/SegmentMetadataClient.java]: To confirm, this is an internal API, so we are not breaking API compatibility with this change, are we?

Comment (tkaitchuck) [reply client/src/main/java/io/pravega/client/segment/impl/SegmentMetadataClient.java]: Yes, this is an internal API

File: test/integration/src/test/java/io/pravega/test/integration/TruncateTest.java (hunk 3 above)
Comment (tkaitchuck) [reply test/integration/src/test/java/io/pravega/test/integration/TruncateTest.java]: Done

File: test/integration/src/test/java/io/pravega/test/integration/TruncateTest.java (hunk 2 above)
Comment (tkaitchuck) [reply test/integration/src/test/java/io/pravega/test/integration/TruncateTest.java]: The test testStreamTruncatedWithPartialEvent I added above covers that.

File: client/src/main/java/io/pravega/client/segment/impl/SegmentMetadataClient.java (hunk 1 above)
Comment (tkaitchuck) [reply client/src/main/java/io/pravega/client/segment/impl/SegmentMetadataClient.java]: Fixed

**Pull request merged**